README.md
<p align="center">
  <a href="https://izypd.com/pwgan">
    <img src="https://cdn.jsdelivr.net/gh/izypd/Gauss@main/PWGAN/PWGAN结果.png">
  </a>
</p>

<h1 align="center">Icon-to-Icon Translation using Pixel-Level Weight Adversarial Networks</h1>

<p align="center">
  <a href="https://izypd.com/pwgan">
    预览
  </a>
  <a href="http://bit.kuas.edu.tw/~jni/2022/vol7/s1/11.JNI0275.pdf">
    论文
  </a>
</p>

<div align="center">

PWGAN 是一种基于生成对抗网络的图标转换方法，与现有方法对比，在定量和定性评估中，都达到了目前最好的水准。

本项目是 PWGAN 的 PyTorch 实现。

助力智能图标设计，只需提供转换目标图标集，与输入图标集共同参与无监督训练。利用训练得到的模型，可自动将输入图标集转换为目标图标集的风格，实现智能图标风格迁移。

[![Python](https://img.shields.io/badge/Python-3.6+-3776AB?style=for-the-badge)](https://www.python.org)
[![PyTorch](https://img.shields.io/badge/PyTorch-1.4+-EE4C2C?style=for-the-badge)](https://pytorch.org)
[![PR Welcome](https://img.shields.io/badge/PR-welcome-60ca2b?style=for-the-badge)](https://github.com/izypd/blog-react/pulls)
[![License](https://img.shields.io/badge/License-GPL-60ca2b?style=for-the-badge)](https://www.gnu.org/licenses/gpl-3.0.html)

</div>

## Prerequisites
- Linux or macOS
- Python 3.6+
- PyTorch 1.4+ with CUDA/ROCm or CPU

## 快速上手

### 文档

- 模型训练、测试 [tips](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/docs/tips.md)

- 常见[问题](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/docs/qa.md)

- 自定义数据集的[template](data/template_dataset.py)

- 自定义模型的[template](models/template_model.py)

- 代码结构[overview](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/docs/overview.md)

- 如何使用自己的数据集[datasets](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/docs/datasets.md)

### 安装

- Clone this repo:
```bash
git clone https://github.com/izypd/PWGAN.git
cd PWGAN
```

- 安装[PyTorch](http://pytorch.org)和其他依赖（如torchvision，[visdom](https://github.com/facebookresearch/visdom)，[dominate](https://github.com/Knio/dominate)）。
```bash
pip install -r requirements.txt
```

### 模型训练、测试
- 自行收集图标数据集(考虑到可能出现的版权问题，不放出论文实验使用的图标数据集，请根据论文中图标集的引用信息，自行收集数据集)
- 想实时查看训练结果和损失图，运行`python -m visdom.server`，然后打开http://localhost:8097。
- 训练模型:
```bash
python train.py --dataroot ./datasets/Cardicons2_MBEStyle --name C2M_square_0.618 --gpu_ids 0 \
--no_dropout --no_flip \
--cycle_use_weighted --identity_use_weighted \
--cycle_center 0.618 --identity_center 0.618 --center_type square
```
要查看更多训练过程中的结果，请打开`./checkpoints/C2M_square_0.618/web/index.html`。
- 测试模型:
```bash
python test.py --dataroot ./datasets/Cardicons2_MBEStyle --name C2M_square_0.618 --gpu_ids 0 \
--no_dropout --no_flip --phase test
```
- 测试结果会保存至`./results/C2M_square_0.618/latest_test/index.html`。

### 计算生成图标集与原图标集间的 Circle weighted quadratic loss

```bash
python weighted_loss.py SOURCE_DIR GENERATED_DIR
```

## 引用

如果你的研究使用了本代码，请引用：
```
@article{PWGAN,
  author    = {Yao-Ming Pan and
               Kai-Biao Lin and
               Chin-Ling Chen},
  title     = {Icon-to-Icon Translation using Pixel-Level Weight Adversarial Networks},
  journal   = {Journal of Network Intelligence},
  volume    = {7},
  number    = {1},
  pages     = {147--160},
  year      = {2022}
}
```

## 致谢
我们的代码，很大程度上启发于[CycleGAN](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix)。


CHANGELOG.md
# 22-03-01

添加 CHANGELOG.md、LICENSE
添加 .gitignore
添加 PWGAN 核心代码
添加 weighted quadratic loss 计算脚本 weighted_loss.py
添加 python 依赖文件 requirements.txt

# 22-03-02

优化 base_options.py 默认值
添加 README.md
修改 README.md


weighted_loss.py
import torch
import torchvision

import math
from pathlib import Path
import typer


def get_square_weight_matrix(n_channels, img_size, center_rate, device):
    m = torch.ones(n_channels, img_size, img_size).to(device)
    half_not_center_size = int(img_size / 2 * (1 - center_rate))
    if half_not_center_size == 0:
        return m
    for i in range(half_not_center_size - 1, -1, -1):
        weight = 1 - (i - half_not_center_size) ** 2 / (half_not_center_size ** 2)
        m[:, i] = weight
        m[:, :, i] = weight
        m[:, img_size - 1 - i] = weight
        m[:, :, img_size - 1 - i] = weight
    return m


def get_circle_weight_matrix(n_channels, img_size, center_rate, device):
    m = torch.ones(n_channels, img_size, img_size).to(device)
    half_not_center_size = int(img_size / 2 * (1 - center_rate))
    dict_x_weight = {}
    for i in range(half_not_center_size - 1, -1, -1):
        weight = 1 - (i - half_not_center_size) ** 2 / (half_not_center_size ** 2)
        dict_x_weight[i] = weight
    for x in range(img_size):
        for y in range(img_size):
            center_i = img_size // 2 - 1
            x_subtract_center = x - center_i
            y_subtract_center = y - center_i
            if x_subtract_center <= 0:
                corresponding_x = -x_subtract_center
            else:
                corresponding_x = x_subtract_center - 1
            if y_subtract_center <= 0:
                corresponding_y = -y_subtract_center
            else:
                corresponding_y = y_subtract_center - 1
            power_sum_sqrt = math.sqrt(corresponding_x ** 2 + corresponding_y ** 2)
            half_size = img_size // 2
            half_center_size = half_size - half_not_center_size
            if power_sum_sqrt >= half_size - 1:
                m[:, x, y] = 0.0
            elif power_sum_sqrt >= half_center_size:
                m[:, x, y] = dict_x_weight[half_size - 1 - int(power_sum_sqrt)]
    return m


def get_diamond_weight_matrix(n_channels, img_size, center_rate, device):
    m = torch.ones(n_channels, img_size, img_size).to(device)
    half_not_center_size = int(img_size / 2 * (1 - center_rate))
    dict_x_weight = {}
    for i in range(half_not_center_size - 1, -1, -1):
        weight = 1 - (i - half_not_center_size) ** 2 / (half_not_center_size ** 2)
        dict_x_weight[i] = weight
    for x in range(img_size):
        for y in range(img_size):
            center_i = img_size // 2 - 1
            x_subtract_center = x - center_i
            y_subtract_center = y - center_i
            if x_subtract_center <= 0:
                corresponding_x = -x_subtract_center
            else:
                corresponding_x = x_subtract_center - 1
            if y_subtract_center <= 0:
                corresponding_y = -y_subtract_center
            else:
                corresponding_y = y_subtract_center - 1
            corresponding_x_plus_y = corresponding_x + corresponding_y
            first_corresponding_x = img_size // 2 - half_not_center_size
            last_corresponding_x = img_size // 2 - 1
            if corresponding_x_plus_y > last_corresponding_x:
                m[:, x, y] = dict_x_weight[0]
            elif corresponding_x_plus_y >= first_corresponding_x:
                m[:, x, y] = dict_x_weight[last_corresponding_x - corresponding_x_plus_y]
    return m


def loss(input_dir: str, output_dir: str, device='cpu', type='circle', size=192, rate=0.5, img_extension='png'):
    if type == 'circle':
        weight_matrix = get_circle_weight_matrix(3, size, rate, device)
    if type == 'square':
        weight_matrix = get_square_weight_matrix(3, size, rate, device)
    if type == 'diamond':
        weight_matrix = get_diamond_weight_matrix(3, size, rate, device)
    img_loss = torch.nn.MSELoss()

    input_img_paths = sorted(Path(input_dir).glob(f'*.{img_extension}'))
    output_img_paths = sorted(Path(output_dir).glob(f'*.{img_extension}'))
    input_output_paths = zip(input_img_paths, output_img_paths)

    all_weighted_loss = 0.0
    for input_output_path in input_output_paths:
        input_path = str(input_output_path[0])
        output_path = str(input_output_path[1])
        input_img = torchvision.io.read_image(input_path).type_as(torch.FloatTensor()).to(device)
        output_img = torchvision.io.read_image(output_path).type_as(torch.FloatTensor()).to(device)

        weighted_loss = img_loss(input_img * weight_matrix, output_img * weight_matrix)
        all_weighted_loss += float(weighted_loss)

    num = len(input_img_paths)
    avg_weighted_loss = all_weighted_loss / num
    print(avg_weighted_loss)


if __name__ == "__main__":
    typer.run(loss)


data/image_folder.py
"""A modified image folder class

We modify the official PyTorch image folder (https://github.com/pytorch/vision/blob/master/torchvision/datasets/folder.py)
so that this class can load images from both current directory and its subdirectories.
"""

import torch.utils.data as data

from PIL import Image
import os

IMG_EXTENSIONS = [
    '.jpg', '.JPG', '.jpeg', '.JPEG',
    '.png', '.PNG', '.ppm', '.PPM', '.bmp', '.BMP',
    '.tif', '.TIF', '.tiff', '.TIFF',
    '.webp', '.WEBP',
]


def is_image_file(filename):
    return any(filename.endswith(extension) for extension in IMG_EXTENSIONS)


def make_dataset(dir, max_dataset_size=float("inf")):
    images = []
    assert os.path.isdir(dir), '%s is not a valid directory' % dir

    for root, _, fnames in sorted(os.walk(dir)):
        for fname in fnames:
            if is_image_file(fname):
                path = os.path.join(root, fname)
                images.append(path)
    return images[:min(max_dataset_size, len(images))]


def default_loader(path):
    return Image.open(path).convert('RGB')


class ImageFolder(data.Dataset):

    def __init__(self, root, transform=None, return_paths=False,
                 loader=default_loader):
        imgs = make_dataset(root)
        if len(imgs) == 0:
            raise(RuntimeError("Found 0 images in: " + root + "\n"
                               "Supported image extensions are: " + ",".join(IMG_EXTENSIONS)))

        self.root = root
        self.imgs = imgs
        self.transform = transform
        self.return_paths = return_paths
        self.loader = loader

    def __getitem__(self, index):
        path = self.imgs[index]
        img = self.loader(path)
        if self.transform is not None:
            img = self.transform(img)
        if self.return_paths:
            return img, path
        else:
            return img

    def __len__(self):
        return len(self.imgs)


models/cycle_gan_model.py
import torch
import itertools
from util.image_pool import ImagePool
from .base_model import BaseModel
from . import networks

import math


class CycleGANModel(BaseModel):
    """
    This class implements the CycleGAN model, for learning image-to-image translation without paired data.

    The model training requires '--dataset_mode unaligned' dataset.
    By default, it uses a '--netG resnet_9blocks' ResNet generator,
    a '--netD basic' discriminator (PatchGAN introduced by pix2pix),
    and a least-square GANs objective ('--gan_mode lsgan').

    CycleGAN paper: https://arxiv.org/pdf/1703.10593.pdf
    """
    @staticmethod
    def modify_commandline_options(parser, is_train=True):
        """Add new dataset-specific options, and rewrite default values for existing options.

        Parameters:
            parser          -- original option parser
            is_train (bool) -- whether training phase or test phase. You can use this flag to add training-specific or test-specific options.

        Returns:
            the modified parser.

        For CycleGAN, in addition to GAN losses, we introduce lambda_A, lambda_B, and lambda_identity for the following losses.
        A (source domain), B (target domain).
        Generators: G_A: A -> B; G_B: B -> A.
        Discriminators: D_A: G_A(A) vs. B; D_B: G_B(B) vs. A.
        Forward cycle loss:  lambda_A * ||G_B(G_A(A)) - A|| (Eqn. (2) in the paper)
        Backward cycle loss: lambda_B * ||G_A(G_B(B)) - B|| (Eqn. (2) in the paper)
        Identity loss (optional): lambda_identity * (||G_A(B) - B|| * lambda_B + ||G_B(A) - A|| * lambda_A) (Sec 5.2 "Photo generation from paintings" in the paper)
        Dropout is not used in the original CycleGAN paper.
        """
        parser.set_defaults(no_dropout=True)  # default CycleGAN did not use dropout
        if is_train:
            parser.add_argument('--lambda_A', type=float, default=10.0, help='weight for cycle loss (A -> B -> A)')
            parser.add_argument('--lambda_B', type=float, default=10.0, help='weight for cycle loss (B -> A -> B)')
            parser.add_argument('--lambda_identity', type=float, default=0.5, help='use identity mapping. Setting lambda_identity other than 0 has an effect of scaling the weight of the identity mapping loss. For example, if the weight of the identity loss should be 10 times smaller than the weight of the reconstruction loss, please set lambda_identity = 0.1')
            parser.add_argument('--cycle_use_weighted', action='store_true', help='use weighted cycle loss')
            parser.add_argument('--identity_use_weighted', action='store_true', help='use weighted identity loss')
            parser.add_argument('--center_type', type=str, default='square',
                                help='type of center part. [square | circle | diamond]')
            parser.add_argument('--cycle_center', type=float, default=0.618, help='proportion of center part when calculate center-weighted cycle loss')
            parser.add_argument('--identity_center', type=float, default=0.618, help='proportion of center part when calculate center-weighted identity loss')

        return parser

    def __init__(self, opt):
        """Initialize the CycleGAN class.

        Parameters:
            opt (Option class)-- stores all the experiment flags; needs to be a subclass of BaseOptions
        """
        BaseModel.__init__(self, opt)
        # specify the training losses you want to print out. The training/test scripts will call <BaseModel.get_current_losses>
        self.loss_names = ['D_A', 'G_A', 'cycle_A', 'idt_A', 'D_B', 'G_B', 'cycle_B', 'idt_B']
        # specify the images you want to save/display. The training/test scripts will call <BaseModel.get_current_visuals>
        visual_names_A = ['real_A', 'fake_B', 'rec_A']
        visual_names_B = ['real_B', 'fake_A', 'rec_B']
        if self.isTrain and self.opt.lambda_identity > 0.0:  # if identity loss is used, we also visualize idt_B=G_A(B) ad idt_A=G_A(B)
            visual_names_A.append('idt_B')
            visual_names_B.append('idt_A')

        self.visual_names = visual_names_A + visual_names_B  # combine visualizations for A and B
        # specify the models you want to save to the disk. The training/test scripts will call <BaseModel.save_networks> and <BaseModel.load_networks>.
        if self.isTrain:
            self.model_names = ['G_A', 'G_B', 'D_A', 'D_B']
        else:  # during test time, only load Gs
            self.model_names = ['G_A', 'G_B']

        # define networks (both Generators and discriminators)
        # The naming is different from those used in the paper.
        # Code (vs. paper): G_A (G), G_B (F), D_A (D_Y), D_B (D_X)
        self.netG_A = networks.define_G(opt.input_nc, opt.output_nc, opt.ngf, opt.netG, opt.norm,
                                        not opt.no_dropout, opt.init_type, opt.init_gain, self.gpu_ids, opt)
        self.netG_B = networks.define_G(opt.output_nc, opt.input_nc, opt.ngf, opt.netG, opt.norm,
                                        not opt.no_dropout, opt.init_type, opt.init_gain, self.gpu_ids, opt)

        if self.isTrain:  # define discriminators
            self.netD_A = networks.define_D(opt.output_nc, opt.ndf, opt.netD,
                                            opt.n_layers_D, opt.norm, opt.init_type, opt.init_gain, self.gpu_ids, opt)
            self.netD_B = networks.define_D(opt.input_nc, opt.ndf, opt.netD,
                                            opt.n_layers_D, opt.norm, opt.init_type, opt.init_gain, self.gpu_ids, opt)

        if self.isTrain:
            if opt.lambda_identity > 0.0:  # only works when input and output images have the same number of channels
                assert(opt.input_nc == opt.output_nc)
            self.fake_A_pool = ImagePool(opt.pool_size)  # create image buffer to store previously generated images
            self.fake_B_pool = ImagePool(opt.pool_size)  # create image buffer to store previously generated images
            # define loss functions
            self.criterionGAN = networks.GANLoss(opt.gan_mode).to(self.device)  # define GAN loss.
            self.criterionCycle = torch.nn.L1Loss()
            self.criterionIdt = torch.nn.L1Loss()
            # initialize optimizers; schedulers will be automatically created by function <BaseModel.setup>.
            self.optimizer_G = torch.optim.Adam(itertools.chain(self.netG_A.parameters(), self.netG_B.parameters()), lr=opt.lr, betas=(opt.beta1, 0.999))
            self.optimizer_D = torch.optim.Adam(itertools.chain(self.netD_A.parameters(), self.netD_B.parameters()), lr=opt.lr, betas=(opt.beta1, 0.999))
            self.optimizers.append(self.optimizer_G)
            self.optimizers.append(self.optimizer_D)

            self.cycle_use_weighted = opt.cycle_use_weighted
            self.identity_use_weighted = opt.identity_use_weighted
            if self.cycle_use_weighted:
                if opt.center_type == 'square':
                    self.cycle_weight_matrix_A = self.get_square_weight_matrix(opt.input_nc, opt.crop_size, opt.cycle_center)
                    self.cycle_weight_matrix_B = self.get_square_weight_matrix(opt.output_nc, opt.crop_size, opt.cycle_center)
                if opt.center_type == 'circle':
                    self.cycle_weight_matrix_A = self.get_circle_weight_matrix(opt.input_nc, opt.crop_size, opt.cycle_center)
                    self.cycle_weight_matrix_B = self.get_circle_weight_matrix(opt.output_nc, opt.crop_size, opt.cycle_center)
                if opt.center_type == 'diamond':
                    self.cycle_weight_matrix_A = self.get_diamond_weight_matrix(opt.input_nc, opt.crop_size, opt.cycle_center)
                    self.cycle_weight_matrix_B = self.get_diamond_weight_matrix(opt.output_nc, opt.crop_size, opt.cycle_center)
            if self.identity_use_weighted:
                if opt.center_type == 'square':
                    self.identity_weight_matrix_A = self.get_square_weight_matrix(opt.input_nc, opt.crop_size, opt.identity_center)
                    self.identity_weight_matrix_B = self.get_square_weight_matrix(opt.output_nc, opt.crop_size, opt.identity_center)
                if opt.center_type == 'circle':
                    self.identity_weight_matrix_A = self.get_circle_weight_matrix(opt.input_nc, opt.crop_size, opt.identity_center)
                    self.identity_weight_matrix_B = self.get_circle_weight_matrix(opt.output_nc, opt.crop_size, opt.identity_center)
                if opt.center_type == 'diamond':
                    self.identity_weight_matrix_A = self.get_diamond_weight_matrix(opt.input_nc, opt.crop_size, opt.identity_center)
                    self.identity_weight_matrix_B = self.get_diamond_weight_matrix(opt.output_nc, opt.crop_size, opt.identity_center)

    def set_input(self, input):
        """Unpack input data from the dataloader and perform necessary pre-processing steps.

        Parameters:
            input (dict): include the data itself and its metadata information.

        The option 'direction' can be used to swap domain A and domain B.
        """
        AtoB = self.opt.direction == 'AtoB'
        self.real_A = input['A' if AtoB else 'B'].to(self.device)
        self.real_B = input['B' if AtoB else 'A'].to(self.device)
        self.image_paths = input['A_paths' if AtoB else 'B_paths']

    def forward(self):
        """Run forward pass; called by both functions <optimize_parameters> and <test>."""
        self.fake_B = self.netG_A(self.real_A)  # G_A(A)
        self.rec_A = self.netG_B(self.fake_B)   # G_B(G_A(A))
        self.fake_A = self.netG_B(self.real_B)  # G_B(B)
        self.rec_B = self.netG_A(self.fake_A)   # G_A(G_B(B))

    def backward_D_basic(self, netD, real, fake):
        """Calculate GAN loss for the discriminator

        Parameters:
            netD (network)      -- the discriminator D
            real (tensor array) -- real images
            fake (tensor array) -- images generated by a generator

        Return the discriminator loss.
        We also call loss_D.backward() to calculate the gradients.
        """
        # Real
        pred_real = netD(real)
        loss_D_real = self.criterionGAN(pred_real, True)
        # Fake
        pred_fake = netD(fake.detach())
        loss_D_fake = self.criterionGAN(pred_fake, False)
        # Combined loss and calculate gradients
        loss_D = (loss_D_real + loss_D_fake) * 0.5
        loss_D.backward()
        return loss_D

    def backward_D_A(self):
        """Calculate GAN loss for discriminator D_A"""
        fake_B = self.fake_B_pool.query(self.fake_B)
        self.loss_D_A = self.backward_D_basic(self.netD_A, self.real_B, fake_B)

    def backward_D_B(self):
        """Calculate GAN loss for discriminator D_B"""
        fake_A = self.fake_A_pool.query(self.fake_A)
        self.loss_D_B = self.backward_D_basic(self.netD_B, self.real_A, fake_A)

    def get_square_weight_matrix(self, n_channels, img_size, center_rate):
        m = torch.ones(n_channels, img_size, img_size).to(self.device)
        half_not_center_size = int(img_size / 2 * (1 - center_rate))
        if half_not_center_size == 0:
            return m
        for i in range(half_not_center_size - 1, -1, -1):
            weight = 1 - (i - half_not_center_size) ** 2 / (half_not_center_size ** 2)
            m[:, i] = weight
            m[:, :, i] = weight
            m[:, img_size - 1 - i] = weight
            m[:, :, img_size - 1 - i] = weight
        return m

    def get_circle_weight_matrix(self, n_channels, img_size, center_rate):
        m = torch.ones(n_channels, img_size, img_size).to(self.device)
        half_not_center_size = int(img_size / 2 * (1 - center_rate))
        dict_x_weight = {}
        for i in range(half_not_center_size - 1, -1, -1):
            weight = 1 - (i - half_not_center_size) ** 2 / (half_not_center_size ** 2)
            dict_x_weight[i] = weight
        for x in range(img_size):
            for y in range(img_size):
                center_i = img_size // 2 - 1
                x_subtract_center = x - center_i
                y_subtract_center = y - center_i
                if x_subtract_center <= 0:
                    corresponding_x = -x_subtract_center
                else:
                    corresponding_x = x_subtract_center - 1
                if y_subtract_center <= 0:
                    corresponding_y = -y_subtract_center
                else:
                    corresponding_y = y_subtract_center - 1
                power_sum_sqrt = math.sqrt(corresponding_x ** 2 + corresponding_y ** 2)
                half_size = img_size // 2
                half_center_size = half_size - half_not_center_size
                if power_sum_sqrt >= half_size - 1:
                    m[:, x, y] = 0.0
                elif power_sum_sqrt >= half_center_size:
                    m[:, x, y] = dict_x_weight[half_size - 1 - int(power_sum_sqrt)]
        return m

    def get_diamond_weight_matrix(self, n_channels, img_size, center_rate):
        m = torch.ones(n_channels, img_size, img_size).to(self.device)
        half_not_center_size = int(img_size / 2 * (1 - center_rate))
        dict_x_weight = {}
        for i in range(half_not_center_size - 1, -1, -1):
            weight = 1 - (i - half_not_center_size) ** 2 / (half_not_center_size ** 2)
            dict_x_weight[i] = weight
        for x in range(img_size):
            for y in range(img_size):
                center_i = img_size // 2 - 1
                x_subtract_center = x - center_i
                y_subtract_center = y - center_i
                if x_subtract_center <= 0:
                    corresponding_x = -x_subtract_center
                else:
                    corresponding_x = x_subtract_center - 1
                if y_subtract_center <= 0:
                    corresponding_y = -y_subtract_center
                else:
                    corresponding_y = y_subtract_center - 1
                corresponding_x_plus_y = corresponding_x + corresponding_y
                first_corresponding_x = img_size // 2 - half_not_center_size
                last_corresponding_x = img_size // 2 - 1
                if corresponding_x_plus_y > last_corresponding_x:
                    m[:, x, y] = dict_x_weight[0]
                elif corresponding_x_plus_y >= first_corresponding_x:
                    m[:, x, y] = dict_x_weight[last_corresponding_x - corresponding_x_plus_y]
        return m

    def backward_G(self):
        """Calculate the loss for generators G_A and G_B"""
        lambda_idt = self.opt.lambda_identity
        lambda_A = self.opt.lambda_A
        lambda_B = self.opt.lambda_B
        # Identity loss
        if lambda_idt > 0:
            # G_A should be identity if real_B is fed: ||G_A(B) - B||
            self.idt_A = self.netG_A(self.real_B)
            if self.identity_use_weighted:
                self.loss_idt_A = self.criterionIdt(self.idt_A * self.identity_weight_matrix_B, self.real_B * self.identity_weight_matrix_B) * lambda_B * lambda_idt
            else:
                self.loss_idt_A = self.criterionIdt(self.idt_A, self.real_B) * lambda_B * lambda_idt
            # G_B should be identity if real_A is fed: ||G_B(A) - A||
            self.idt_B = self.netG_B(self.real_A)
            if self.identity_use_weighted:
                self.loss_idt_B = self.criterionIdt(self.idt_B * self.identity_weight_matrix_A, self.real_A * self.identity_weight_matrix_A) * lambda_A * lambda_idt
            else:
                self.loss_idt_B = self.criterionIdt(self.idt_B, self.real_A) * lambda_A * lambda_idt
        else:
            self.loss_idt_A = 0
            self.loss_idt_B = 0

        # GAN loss D_A(G_A(A))
        self.loss_G_A = self.criterionGAN(self.netD_A(self.fake_B), True)
        # GAN loss D_B(G_B(B))
        self.loss_G_B = self.criterionGAN(self.netD_B(self.fake_A), True)
        # Forward cycle loss || G_B(G_A(A)) - A||
        if self.cycle_use_weighted:
            self.loss_cycle_A = self.criterionCycle(self.rec_A * self.cycle_weight_matrix_A, self.real_A * self.cycle_weight_matrix_A) * lambda_A
        else:
            self.loss_cycle_A = self.criterionCycle(self.rec_A, self.real_A) * lambda_A
        # Backward cycle loss || G_A(G_B(B)) - B||
        if self.cycle_use_weighted:
            self.loss_cycle_B = self.criterionCycle(self.rec_B * self.cycle_weight_matrix_B, self.real_B * self.cycle_weight_matrix_B) * lambda_B
        else:
            self.loss_cycle_B = self.criterionCycle(self.rec_B, self.real_B) * lambda_B
        # combined loss and calculate gradients
        self.loss_G = self.loss_G_A + self.loss_G_B + self.loss_cycle_A + self.loss_cycle_B + self.loss_idt_A + self.loss_idt_B
        self.loss_G.backward()

    def optimize_parameters(self):
        """Calculate losses, gradients, and update network weights; called in every training iteration"""
        # forward
        self.forward()      # compute fake images and reconstruction images.
        # G_A and G_B
        self.set_requires_grad([self.netD_A, self.netD_B], False)  # Ds require no gradients when optimizing Gs
        self.optimizer_G.zero_grad()  # set G_A and G_B's gradients to zero
        self.backward_G()             # calculate gradients for G_A and G_B
        self.optimizer_G.step()       # update G_A and G_B's weights
        # D_A and D_B
        self.set_requires_grad([self.netD_A, self.netD_B], True)
        self.optimizer_D.zero_grad()   # set D_A and D_B's gradients to zero
        self.backward_D_A()      # calculate gradients for D_A
        self.backward_D_B()      # calculate graidents for D_B
        self.optimizer_D.step()  # update D_A and D_B's weights
